doc/specs/#885 - winrt Terminal Settings.md

```diff
@@ -0,0 +1,205 @@
+ within the `Microsoft.Terminal.Settings.Model` namespace. This allows them to be interacted with across
+ different project layers.
+
+Additionally, the top-level object holding all of the settings related to Windows Terminal will be
+`AppSettings`. The runtime class will look something like this:
+```c++
+runtimeclass AppSettings
+{
+    // global settings that generally only affect TerminalApp
+    GlobalSettings Globals { get; };
+
+    // the list of profiles available to Windows Terminal (custom and dynamically generated)
+    IObservableVector<Profile> Profiles { get; };
+
+    // the list of key bindings indexed by their unique key chord
+    IObservableMap<KeyChord, Action> Keybindings { get; };
+
+    // the list of command palette commands indexed by their name (custom or auto-generated)
+    IObservableMap<String, Action> Commands { get; };
+
+    // the list of color schemes indexed by their unique names
+    IObservableMap<String, ColorScheme> Schemes { get; };
+
+    // a list of warnings encountered during the serialization of the JSON
+    IVector<SerializationWarnings> Warnings { get; };
+}
+```
+
+Introducing a number of observable WinRT types allows for other components to subscribe to changes
+ to these collections. A particular example of this being used can be a preview of the TermControl in
+ the upcoming Settings UI.
+
+Adjacent to the introduction of `AppSettings`, `IControlSettings` and `ICoreSettings` will be moved
+ to the `Microsoft.Terminal.TerminalControl` namespace. This allows for a better consumption of the
+ settings model that is covered later in the (Consumption section)[#terminal-settings-model:-consumption].
+
+
+### Terminal Settings Model: Serialization and Deserialization
+
+Introducing these `Microsoft.Terminal.Settings.Model` WinRT objects also allow the serialization and deserialization
+ logic from TerminalApp to be moved to TerminalSettings. `JsonUtils` introduces several quick and easy methods
+ for setting serialization. This will be moved into the `Microsoft.Terminal.Settings.Model` namespace too.
+
+Deserialization will be an extension of the existing `JsonUtils` `ConversionTrait` struct template. `ConversionTrait`
+ already includes `FromJson` and `CanConvert`. Deserialization would be handled by a `ToJson` function.
+
+
+### Interacting with the Terminal Settings Model
+
+Separate projects can interact with `AppSettings` to extract data, subscribe to changes, and commit changes.
+This API will look something like this:
+```c++
+runtimeclass AppSettings
+{
+    AppSettings();
+
+    // Load a settings file, and layer those changes on top of the existing AppSettings
+    void LayerSettings(String path);
+
+    // Create a copy of the existing AppSettings
+    AppSettings Clone();
+    
+    // Compares object to "source" and applies changes to
+    // the settings file at "outPath"
+    void ApplyChanges(String outPath, AppSettings source);

```

Comment:
Why is this "apply changes from source object" and not just "save"? Instead of it being `original.Apply(filename, copy)` why would it not just be `copy.Save(filename)`?

Reply:
Included my train of thought below. Then realized as I was explaining it that it's not necessary. Updating to the `Save()` API instead.

<details>
Ah. Yes. So if we go the `Save(String path)` route, I see it working something like this:
- Construct `AppSettings` from `defaults.json` and `settings.json` (again)
- Compare each AppSettings field between `copy` and `original`
- if they differ, deserialize `copy`'s setting and inject it into `original`

That first step is us really just setting up to compare to `original`. So, I figured we could save some time and have a reference to `source`.

Then again, we literally have a reference to `source`. So, we don't actually have to construct it.
</details>